core/payments/ondemand/on_demand_payment_validator.go

```diff
@@ -0,0 +1,179 @@
+package ondemand
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/Layr-Labs/eigenda/core/meterer"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/hashicorp/go-multierror"
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+// OnDemandPaymentValidator validates on-demand payments for multiple accounts
+type OnDemandPaymentValidator struct {
+	logger logging.Logger
+	// A cache of the ledgers being tracked.
+	//
+	// New OnDemandLedger entries are added to this cache as Debit requests are received from new accounts. Least
+	// recently used OnDemandLedger entries are removed if the cache gets above the configured size. Since on-demand
+	// payment data is stored in a persistent way, deleting an OnDemandLedger from memory doesn't result in data loss:
+	// it just means that a new OnDemandLedger object will need to be constructed if any future Debits must be handled
+	// from that account.
+	ledgers *lru.Cache[gethcommon.Address, *OnDemandLedger]
+	// protects concurrent access to the ledgers cache during ledger creation
+	ledgerCreationLock sync.Mutex
```

Comment:
https://github.com/hashicorp/golang-lru says its thread-safe, are you sure we need this mutex?

Reply:
The structure itself is threadsafe, but this lock isn't to protect access to the struct.

Consider 2 callers fail to access a ledger via the fast path, and enter the slow path at the same moment. They will both construct a new ledger, add it to the cache, and then return it for use.

That means that whichever caller finishes construction first will return a ledger that gets used for the dispersal, but that instance will be overwritten in the cache by the second caller, effectively making the first dispersal "free"

Reply:
Are you sure that's how it works? Wouldn't LRU instead add both callers' ledgers to the cache and evict too other ledgers that haven't been used in a while?